recipes/lora_finetune_distributed_td.py

```diff
@@ -0,0 +1,1119 @@
+        self,
+        cfg_dataloader: DictConfig,
+        batch_size: int,
+    ) -> BaseNode:
+        """
+        All data related setup happens here. Currently this recipe only supports
+        Map-style Datasets which fit into memory and an option for random shuffling.
+        Samplers, iterable datasets, and streaming datasets are not supported.
+        """
+        world_size, rank = training.get_world_size_and_rank()
+
+        # Get global settings
+        shuffle = cfg_dataloader.shuffle
+        parallel_method = cfg_dataloader.get("parallel_method", "thread")
+        packed = cfg_dataloader.get("packed", False)
+        streaming = cfg_dataloader.get("streaming", False)
+        num_workers = cfg_dataloader.get("num_workers", 0)
+        pin_memory = cfg_dataloader.get("pin_memory", True)
+        collate_fn = cfg_dataloader.collate_fn
+
+        cfg_datasets = cfg_dataloader["datasets"]
+        weights, datasets = {}, {}
+        if isinstance(cfg_datasets, DictConfig):
+            for k, cfg_and_weight in cfg_datasets.items():
+                weights[k] = float(cfg_and_weight.get("weight", 1.0))
+                datasets[k] = Prefetcher(
+                    self._setup_one_dataset(
+                        cfg_dataset=cfg_and_weight.dataset,
+                        global_shuffle=shuffle,
+                        global_parallel_method=parallel_method,
+                        global_streaming=streaming,
+                        global_num_workers=num_workers,
+                    ),
+                    prefetch_factor=8,
+                )
+
+        # Instantiate collate_fn
+        if "left_pad_sequence" in collate_fn:
+            raise RuntimeError("left_pad_sequence collator is only for inference.")
+        collate_fn = _get_component_from_path(collate_fn)
+
+        # TODO: add multi-dataset mixer
+        if num_workers == 0:
+            _Mapper = Mapper
+        else:
+            _Mapper = partial(
+                ParallelMapper,
+                num_workers=num_workers,
+                method=parallel_method,
+            )
+        if len(cfg_datasets) == 1:
+            node = next(iter(datasets.values()))  # TODO: multi dataset
+        else:
+            node = MultiDatasetWeightedSampler(
+                source_nodes=datasets,
+                weights=weights,
+            )
+        node = Batcher(node, batch_size, drop_last=True)
+        node = _Mapper(
+            node,
+            map_fn=(
+                partial(
+                    collate_fn,  # noqa
+                    padding_idx=self._tokenizer.pad_id,
+                    ignore_idx=self._loss_fn.ignore_index,
+                )
+                if not packed
+                else padded_collate_packed
+            ),
+        )
+        if pin_memory:
+            node = PinMemory(node)
+        if num_workers > 0:
+            node = Prefetcher(node, 2)
+
+        log.info("TorchData nodes are initialized")
+
+        return node
```

Comment:
node feels a bit weird, since we do: self._dataloader = self._setup_dataloader(...)

Should we rename it to dataloader?